x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -0,0 +1,414 @@
+                pos += read;
+                off += read;
+                len -= read;
+            } finally {
+                cacheEntry.decRef();
+            }
+        }
+    }
+
+    /**
+     * Mark the given {@link CacheEntry} as evicted: no new read or write operation can be executed on the file on disk, which will be
+     * deleted from disk once all on-going reads are processed.
+     */
+    private void markAsEvicted(final CacheEntry cacheEntry) {
+        logger.trace(() -> new ParameterizedMessage("marking cache entry [{}] as evicted", cacheEntry));
+        cacheEntry.markAsEvicted();
+    }
+
+    /**
+     * Remove from cache all entries that match the given predicate.
+     *
+     * @param predicate the predicate to evaluate
+     */
+    public void removeFromCache(final Predicate<String> predicate) {
+        if (invalidateOnShutdown) {
+            for (String cacheKey : cache.keys()) {
+                if (predicate.test(cacheKey)) {
+                    cache.invalidate(cacheKey);
+                }
+            }
+        }
+        cache.refresh();
+    }
+
+    /**
+     * Computes the cache key associated to the given Lucene cached file
+     *
+     * @param cacheFile the cached file
+     * @return the cache key
+     */
+    private static String toCacheKey(final Path cacheFile) {
+        return cacheFile.toAbsolutePath().toString();
+    }
+
+    private class CacheEntry implements RefCounted {
+
+        private final CheckedSupplier<IndexInput, IOException> source;
+        private final SparseFileTracker tracker;
+        private final FileChannel channel;
+        private final int sizeOfRange;
+        private final Path path;
+
+        private final AbstractRefCounted refCounter;
+        private volatile boolean evicted;
+
+        CacheEntry(Path path, long length, CheckedSupplier<IndexInput, IOException> source, FileChannel channel, int sizeOfRange) {
+            this.tracker = new SparseFileTracker(path.toString(), length);
+            this.source = Objects.requireNonNull(source);
+            this.channel = Objects.requireNonNull(channel);
+            this.path = Objects.requireNonNull(path);
+            this.sizeOfRange = sizeOfRange;
+            this.refCounter = new AbstractRefCounted(path.toString()) {
+                @Override
+                protected void closeInternal() {
+                    assert refCount() == 0;
+                    IOUtils.closeWhileHandlingException(CacheEntry.this.channel);
+                    assert evicted || invalidateOnShutdown == false;
+                    if (evicted) {
+                        logger.trace(() -> new ParameterizedMessage("deleting cache file [{}]", path));
+                        IOUtils.deleteFilesIgnoringExceptions(path);
+                    }
+                }
+            };
+            this.evicted = false;
+        }
+
+        private long estimateWeight() {
```

Comment:
> I think for now it might be safer to use the length of the file.

I agree, I pushed [this change](https://github.com/elastic/elasticsearch/pull/50693/files#diff-a0405cffc7e6de74715aa3935ac736ecR63)